File: longest-common-subsequence/lcs.rb
def lcs(str1, str2)
  m, n = str1.length, str2.length
  dp = Array.new(m + 1) { Array.new(n + 1, 0) }
  
  (1..m).each do |i|
    (1..n).each do |j|
      if str1[i-1] == str2[j-1]
        dp[i][j] = dp[i-1][j-1] + 1
      else
        dp[i][j] = [dp[i-1][j], dp[i][j-1]].max
      end
    end
  end
  
  dp[m][n]
end

puts lcs("ABCDGH", "AEDFHR")

File: longest-common-subsequence/lcs.clj
(defn lcs [str1 str2]
  (let [m (count str1)
        n (count str2)
        dp (vec (for [i (range (inc m))]
                 (vec (repeat (inc n) 0))))]
    (reduce (fn [dp i]
             (reduce (fn [dp j]
                      (if (= (nth str1 (dec i)) (nth str2 (dec j)))
                        (assoc-in dp [i j] (inc (get-in dp [(dec i) (dec j)])))
                        (assoc-in dp [i j] (max (get-in dp [(dec i) j])
                                               (get-in dp [i (dec j)])))))
                    dp (range 1 (inc n))))
           dp (range 1 (inc m)))
    (get-in dp [m n])))

(println (lcs "ABCDGH" "AEDFHR"))